pset/pset.go

```diff
@@ -283,6 +289,200 @@ func (p *Pset) SanityCheck() error {
 	return nil
 }
 
+func (p *Pset) ValidateAllSignatures() (bool, error) {
+	for i := range p.Inputs {
+		valid, err := p.ValidateInputSignatures(i)
+		if err != nil {
+			return false, err
+		}
+		if !valid {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (p *Pset) ValidateInputSignatures(inputIndex int) (bool, error) {
+	for _, partialSig := range p.Inputs[inputIndex].PartialSigs {
+		valid, err := p.validatePartialSignature(inputIndex, partialSig)
+		if err != nil {
+			return false, err
+		}
+		if !valid {
+			return false, nil
+		}
+	}
+	return true, nil
+}
+
+func (p *Pset) validatePartialSignature(
+	inputIndex int,
+	partialSignature *psbt.PartialSig,
+) (bool, error) {
+	if partialSignature.PubKey == nil {
+		return false, errors.New("no pub key for partial signature")
+	}
+
+	signatureLen := len(partialSignature.Signature)
+	sigHashType := partialSignature.Signature[signatureLen-1]
+	signatureDer := partialSignature.Signature[:signatureLen-1]
+
+	sigHash, script, err := p.getHashAndScriptForSignature(
+		inputIndex,
+		uint32(sigHashType),
+	)
+	if err != nil {
+		return false, err
+	}
+
+	//TODO: check if this is valid only for P2PKH and P2WPKH since we
+	//should not be able to compare script hash with public key hash
+	valid, err := p.verifyScripForPubKey(script, partialSignature.PubKey)
+	if err != nil {
+		return false, err
+	}
+	if !valid {
+		return false, nil
+	}
+
+	pSig, err := btcec.ParseDERSignature(signatureDer, btcec.S256())
+	if err != nil {
+		return false, nil
+	}
+
+	pubKey, err := btcec.ParsePubKey(partialSignature.PubKey, btcec.S256())
+	if err != nil {
+		return false, nil
+	}
+
+	return pSig.Verify(sigHash[:], pubKey), nil
+}
+
+func (p *Pset) getHashAndScriptForSignature(inputIndex int, sigHashType uint32) (
+	[]byte,
+	[]byte,
+	error,
+) {
+	var hash [32]byte
+	var script []byte
+
+	input := p.Inputs[inputIndex]
+
+	if input.NonWitnessUtxo != nil {
+		prevoutHash := p.UnsignedTx.Inputs[inputIndex].Hash
+		utxoHash := input.NonWitnessUtxo.WitnessHash()
+
+		if bytes.Compare(prevoutHash, utxoHash.CloneBytes()) == 1 {
+			return nil, nil,
+				errors.New("mon-witness utxo hash for input doesnt match the " +
+					"hash specified in the prevout")
+		}
+
+		prevoutIndex := p.UnsignedTx.Inputs[inputIndex].Index
+		prevout := input.NonWitnessUtxo.Outputs[prevoutIndex]
+		if input.RedeemScript != nil {
+			script = input.RedeemScript
+		} else {
+			script = prevout.Script
+		}
+
+		switch address.GetScriptType(script) {
+
+		case address.P2WshScript:
+			hash = p.UnsignedTx.HashForWitnessV0(
+				inputIndex,
+				input.WitnessScript,
+				prevout.Value,
+				txscript.SigHashType(sigHashType),
+			)
+			script = input.WitnessScript
+
+		case address.P2WpkhScript:
+			pay, err := payment.FromScript(
+				script,
+				&network.Liquid,
```

Comment:
Is this FromScript used as a "validation" check for the given script? this script == pay.Script (at line 411?) 
Because otherwise we need to manage the network, provably inherit it from Pset struct (which is missing a network field)

Reply:
network is useless here because we don't want an address but just the script, so it's just fine to pass it hardcoded

Reply:
this step actually creates the signing legacy pkh script `OP_DUP OP_HASH160 hash160(pk) OP_EQUALVERIFY OP_CHECKSIG` from the segwit version script `00 hash160(pk)`

Reply:
Done